### DISCUSSION TITLE
Session support

### DISCUSSION THREAD

**Discussion by Robert-M-Muench** [helix-editor/helix#899] [Ideas] [answered]
When I have a number of files open, windows split, etc. it would be nice to save all this as a session, that I can reload with one action.

**Comment by kirawi** [comment, marked as answer]
#401 It's complicated because it has the same fundamental unsafety as multithreading but even worse as it's on the process level. It'll get implemented eventually, but it's not a high priority right now.

**Reply by Robert-M-Muench** [reply]
I am not certain that I understand how these things are related.

I mean HX already knows the file buffer list, why not save it to a simple state JSON file with the session name as key and file-buffer, window-setting, etc. keys per session?

**Reply by kirawi** [reply]
I'm a bit lazy right now, but maybe these issues will help?
https://github.com/neovim/neovim/issues/999
https://github.com/neovim/neovim/pull/2506

**Reply by archseer** [reply]
It's planned in #401 but conflict resolution gets a bit complicated when there's multiple editor instances running all at once and they're all trying to write to the state file.

**Reply by Robert-M-Muench** [reply]
This all ready very complicated which, I think, is not necessary, or maybe I'm thinking about something different.

I have a named session like *my-super-tool* which captures all the things that need to be persistent to restore the session. There are now several simple ways how to do it:

1. write a file per session, named as the session (might need some file-system normalization)
2. write all sessions into one session.json file, each with its section (you can use the simple lock-file approach for multiple instance handling, nothing fancy)

**Reply by kirawi** [reply]
Yes that could be done, but I feel it would incur significant technical debt that would hinder development. Since Helix is OSS, it has the option of doing things at its own pace and maintain a high-quality codebase, which I think will be a better long-term strategy. Your proposal isn't a bad one, but I feel that there's already so much needing to be juggled.

**Reply by kirawi** [reply]
^ Edited the message, realized that it came across differently than I intended.
